File: day3/day3.js
const stdin = process.openStdin();
let data = "";
stdin.on("data", function (chunk) {
    data += chunk;
});

function prio(char) {
    if (char.charCodeAt(0) >= 97) {
        return char.charCodeAt(0) - 96;
    } else {
        return char.charCodeAt(0) - 65 + 27;
    }
}
['a', 'b', 'z', 'A', 'B', 'Z'].forEach((char) => {
    console.log(char, prio(char));
});

stdin.on("end", function () {
    const rucksacks = data
        .trim()
        .split("\n")
        .map((line) => {
            const a = line.substring(0, line.length / 2)
            const b = line.substring(line.length / 2)
            const e = a.split('').find(i => b.includes(i))
            return ({ a, b, e, p: prio(e) });
        })
    console.log({ rucksacks });

    const part1 = rucksacks.reduce((acc, { p }) => acc + p, 0);
    console.log('part 1', part1);

    const groups = data
        .trim()
        .split("\n").reduce((acc, l, i) => {
            if (i % 3 === 0) {
                acc.push([]);
            }
            acc[acc.length - 1].push(l);
            return acc;
        }, []).map(gr => {
            const b = gr[0].split('').find(i => gr[1].includes(i) && gr[2].includes(i));
            return ({ gr, b, p: prio(b) });
        })
    console.log({ groups });

    const part2 = groups.reduce((acc, { p }) => acc + p, 0);
    console.log('part 2', part2);
});

File: day3/day3.exs
backpacks =
  IO.read(:stdio, :all)
  |> String.trim()
  |> String.split("\n")

prio = fn char ->
  code = char |> String.to_charlist() |> hd

  if code >= 97 do
    code - 96
  else
    code - 65 + 27
  end
end

part1 =
  backpacks
  |> Enum.map(fn bp ->
    l = String.length(bp)

    [
      String.slice(bp, 0, (l / 2) |> trunc),
      String.slice(bp, (l / 2) |> trunc, l)
    ]
  end)
  |> Enum.map(fn backpack ->
    other = Enum.at(backpack, 1)
    items = Enum.at(backpack, 0) |> String.split("", trim: true)

    Enum.find(items, fn item ->
      String.contains?(other, item)
    end)
    |> prio.()
  end)
  |> Enum.sum()

IO.puts("Part 1: #{part1}")

part2 =
  backpacks
  |> Enum.chunk_every(3)
  |> Enum.map(fn group ->
    other1 = Enum.at(group, 1)
    other2 = Enum.at(group, 2)
    items = Enum.at(group, 0) |> String.split("", trim: true)

    Enum.find(items, fn item ->
      String.contains?(other1, item) && String.contains?(other2, item)
    end)
    |> prio.()
  end)
  |> Enum.sum()

IO.puts("Part 2: #{part2}")
